### PR TITLE
Fix Linux thread tampering in non-dumpable processes

### PR DESCRIPTION
**Pull request by josiahyan**
Fixes frida#330 partially and frida#319.

### CODE REVIEW

**Comment by josiahyan** [comment]
The test suite is currently failing on Linux x64, although I don't believe my change caused that.
On the commit before this one (27f8e83d41936e087692ada059402e051522574e) I get the error
```
/Core/Process/process_threads: **
Frida:ERROR:../../../frida-gum/tests/core/process.c:156:test_process_process_threads: assertion failed (ctx.number_of_calls >= 2): (1 >= 2)
```

On this commit, I get the same error.

The Stalker test suite is currently passing on Android aarch64, which exercises the code I've changed.

Manually tested this change by attaching to a multi-threaded process on Android aarch64, and using `Process.enumerateThreads()`, which needs to ptrace() other threads to obtain their context. An empty result was observed in the previous commit, while some thread information was dumped after applying this patch.

File: gum/backend-linux/gumprocess-linux.c [hunk 1]
```diff
@@ -381,6 +383,16 @@ gum_process_modify_thread (GumThreadId thread_id,
     desc = tls;
 #endif
 
+    /* Some systems (notably Android on release applications) spawn processes as
```
Comment (oleavr) [inline gum/backend-linux/gumprocess-linux.c]:
```suggestion
    /*
     * Some systems (notably Android on release applications) spawn processes as
```

File: gum/backend-linux/gumprocess-linux.c [hunk 2]
```diff
@@ -381,6 +383,16 @@ gum_process_modify_thread (GumThreadId thread_id,
...
+     * not dumpable by default, disabling ptrace() on that process for anyone
+     * other than root.
+     *
+     * To allow our child to ptrace() this process, we enable this temporarily.
+     */
+    prev_dumpable = prctl (PR_GET_DUMPABLE);
+    if (!(prev_dumpable < 0) && prev_dumpable != 1)
```
Comment (oleavr) [inline gum/backend-linux/gumprocess-linux.c]:
```suggestion
    if (prev_dumpable != -1 && prev_dumpable != 1)
```

File: gum/backend-linux/gumprocess-linux.c [hunk 3]
```diff
@@ -440,6 +452,9 @@ gum_process_modify_thread (GumThreadId thread_id,
       }
     }
 
+    if (!(prev_dumpable < 0) && prev_dumpable != 1)
```
Comment (oleavr) [inline gum/backend-linux/gumprocess-linux.c]:
```suggestion
    if (prev_dumpable != -1 && prev_dumpable != 1)
```

File: gum/backend-linux/gumprocess-linux.c [hunk 4]
```diff
@@ -440,6 +452,9 @@ gum_process_modify_thread (GumThreadId thread_id,
...
+      g_assert (!prctl (PR_SET_DUMPABLE, prev_dumpable));
```
Comment (oleavr) [inline gum/backend-linux/gumprocess-linux.c]: Code with important side-effects should never be inside assertions – which are actually disabled for release builds of Frida:

```suggestion
    {
      int res = prctl (PR_SET_DUMPABLE, prev_dumpable);
      g_assert (res == 0);
    }
```

**Comment by oleavr** [comment]
Actually let me make these tweaks so we can get this released soon.

**Comment by oleavr** [comment]
Tweaked version landed as 73bf271a4e2d4dc2f2e37b3a04bb3681e8160dd0. Thanks!

**Pull request closed without merging**